airqo-platform/AirQo-api
# Hotfix some validations

# Address concerns with some input validations

**_WHAT DOES THIS PR DO?_**

- [x] Address concerns with some input validations (site creation GPS coords, height in deploy activity, site/device in adding events)
- [x] Fixes a bug in deployment, one should not be able to deploy more than once.
- [x] Changes the response format of the bad request errors.
- [x] In case the counter document is missing, create more descriptive error response
- [x] Corrects some HTTP status codes.
- [ ] Query Events by device_number

**_WHAT JIRA STORY/TASK IS RELATED TO THIS PR?_**
N/A

**_WHAT IS THE LINK TO THE PR BRANCH_**
https://github.com/airqo-platform/AirQo-api/tree/hotfix-validations

**_HOW DO I TEST OUT THIS PR?_**
Check out the run instructions for this PR in the READMe file. Run in the dev environment to avoid any chaos.

**_WHICH ENDPOINTS SHOULD BE READY FOR TESTING?:_** 

- [x]  Deploy device

[POST] http://localhost:3000/api/v1/devices/activities/deploy?tenant={TENANT}&deviceName={DEVICE_NAME}
```
{
    "date": "2021-08-05T13:29:16.230Z",
    "height": 3.45,
    "mountType": "Wall",
    "powerType": "solar",
    "latitude": 0.439663534534,
    "longitude": 32.65466733563,
    "isPrimaryInLocation": false,
    "site_id": "{OBJECT_ID}",
    "isUsedForCollocation":false
}

```

- [x]  Add Events 

[POST] http://localhost:3000/api/v1/devices/events?tenant={TENANT}
```
[
    {
        "frequency": "raw",
        "time": "2021-07-28T09:32:42Z",
        "device": "{DEVICE_NAME}",
        "device_id": "{OBJECT_ID}",
        "site_id": "{OBJECT_ID}",
        "device_number": 870144232452
    },
    {
        "frequency": "raw",
        "time": "2021-07-28T09:33:14Z",
        "device": "aq_94",
        "device_id": "{OBJECT_ID}",
        "site_id": "{OBJECT_ID}",
        "device_number": 87014523423423
    }
]
```

- [x]  Create Site (input GPS coords that are more than 5 decimal places)

[POST]  http://localhost:3000/api/v1/devices/sites?tenant={TENANT}
```
{
    "latitude": 0.439654547,
    "longitude": 32.65466,
    "name": "testing",
    "description":"testing"
}
```

- [x] Get Events
[GET] http://localhost:3000/api/v1/devices/events?tenant={TENANT}&recent={YES_or_NO}&device={DEVICE_NAME}

**_IS THERE ANY JENKINS CONSOLE LINK FOR CODE COVERAGE AND BUILD INFO?_**
N/A

**_ARE THERE ANY RELATED PRs?_**
N/A
